man/readpe-ng.1
.TH READPE-NG "1" "October 2026" "objtools 1.10" "User Commands"
.SH NAME
readpe-ng: display information about PE files.

.SH SYNOPSIS
readpe-ng [\fB\-h\fR|\fB\-\-help\fR]
          [\fB\-v\fR|\fB\-\-version\fR]
          [\fB\-V\fR|\fB\-\-verbose\fR]
          [\fB\-S\fR|\fB\-\-section\-headers\fR]
          [\fB\-F\fR|\fB\-\-version\-info\fR]
          [\fB\-a\fR|\fB\-\-all\fR]
          [\fB\-l\fR|\fB\-\-lite\fR]
          [\fB\-H\fR|\fB\-\-headers\fR]
          [\fB\-G\fR|\fB\-\-section\-groups\fR]
          [\fB\-D\fR|\fB\-\-dos\-header\fR]
          [\fB\-O\fR|\fB\-\-dos\-stub\fR]
          [\fB\-N\fR|\fB\-\-nt-header\fR]
          [\fB\-R\fR|\fB\-\-relocs\fR]
          [\fB\-E\fR|\fB\-\-eat\fR]
          [\fB\-I\fR|\fB\-\-iat\fR]
          [\fB\-m\fR|\fB\-\-runtime\fR]
          [\fB\-g\fR|\fB\-\-debug\fR]
          [\fB\-f\fR|\fB\-\-config\fR]
          [\fB\-o\fR|\fB\-\-resource\fR]
          [\fB\-d\fR|\fB\-\-detect\fR]
          [\fB\-x\fR <number or name>|\fB\-\-hex\-dump=\fR<number or name>]
          [\fB\-c\fR <number or name>|\fB\-\-code\-dump=\fR<number or name>]
          [\fB\-s\fR <number or name>|\fB\-\-string\-dump=\fR<number or name>]
          [\fB\-u\fR <number or name>|\fB\-\-unicode\-dump=\fR<number or name>]
          [\fB\-r\fR <number or name>|\fB\-\-relocated\-dump=\fR<number or name>]
          [\fB\-C\fR <number or name>|\fB\-\-disassemble=\fR<number or name>]
          [\fB\-i\fR|\fB\-\-info\fR]
          [\fB\-X\fR|\fB\-\-hash\fR]
          [\fB\-e\fR|\fB\-\-entropy\fR]
          [\fB\-M\fR <options>|\fB\-\-disassembler\-options=\fR<options>]
          [\fB\-t\fR <options>|\fB\-\-script=\fR<options>]
          \fIpefile\fR...

.SH DESCRIPTION
\fBreadpe-ng\fR displays information about one or more PE format object files. The options control what particular information to display.

\fIpefile...\fR are the object files to be examined. 32-bit and 64-bit PE files are supported.

.SH OPTIONS
.IP "\fB\-h\fR" 4
.IX Item "-h"
.PD 0
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
.PD
Print a summary of the options to readpe-ng and exit.
.IP "\fB\-v\fR" 4
.IX Item "-v"
.PD 0
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
.PD
Print the version number of readpe-ng and exit.
.IP "\fB\-V\fR" 4
.IX Item "-V"
.PD 0
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
.PD
Print additional information.
.IP "\fB\-S\fR" 4
.IX Item "-S"
.PD 0
.IP "\fB\-\-section\-headers\fR" 4
.IX Item "--section\-headers"
.PD
Displays the information contained in the file's section headers.
.IP "\fB\-F\fR" 4
.IX Item "-F"
.PD 0
.IP "\fB\-\-version\-info\fR" 4
.IX Item "--version\-info"
.PD
Displays the information contained in the file's version section.
.IP "\fB\-a\fR" 4
.IX Item "-a"
.PD 0
.IP "\fB\-\-all\fR" 4
.IX Item "--all"
.PD
Equivalent to specifying: \fB\-\-section\-headers\fR, \fB\-\-section\-groups\fR, \fB\-\-dos\-header\fR, \fB\-\-dos\-stub\fR, \fB\-\-nt\-header\fR, \fB\-\-version\-info\fR, \fB\-\-resource\fR, \fB\-\-debug\fR, \fB\-\-relocs\fR, \fB\-\-eat\fR, \fB\-\-iat\fR, \fB\-\-runtime\fR and \fB\-\-detect\fR.
.IP "\fB\-l\fR" 4
.IX Item "-l"
.PD 0
.IP "\fB\-\-lite\fR" 4
.IX Item "--lite"
.PD
Equivalent to specifying: \fB\-\-section\-headers\fR, \fB\-\-section\-groups\fR, \fB\-\-dos\-header\fR, \fB\-\-nt\-header\fR, \fB\-\-eat\fR and \fB\-\-iat\fR.
.IP "\fB\-H\fR" 4
.IX Item "-H"
.PD 0
.IP "\fB\-\-headers\fR" 4
.IX Item "--headers"
.PD
Displays the information contained in the file's headers.
Equivalent to specifying: \fB\-\-section\-headers\fR, \fB\-\-dos\-header\fR, and \fB\-\-nt\-header\fR.
.IP "\fB\-G\fR" 4
.IX Item "-G"
.PD 0
.IP "\fB\-\-section\-groups\fR" 4
.IX Item "--section-groups"
.PD
Displays the information contained in the file's version section.
.IP "\fB\-D\fR" 4
.IX Item "-D"
.PD 0
.IP "\fB\-\-dos\-header\fR" 4
.IX Item "--dos-header"
.PD
Displays the contents of the DOS header.
.IP "\fB\-O\fR" 4
.IX Item "-O"
.PD 0
.IP "\fB\-\-dos\-stub\fR" 4
.IX Item "--dos-stub"
.PD
Displays the contents of the DOS stub.
.IP "\fB\-N\fR" 4
.IX Item "-N"
.PD 0
.IP "\fB\-\-nt\-header\fR" 4
.IX Item "--nt-header"
.PD
Displays the contents of the NT header.
.IP "\fB\-R\fR" 4
.IX Item "-R"
.PD 0
.IP "\fB\-\-relocs\fR" 4
.IX Item "--relocs"
.PD
Displays the contents of the file's relocation section.
.IP "\fB\-E\fR" 4
.IX Item "-E"
.PD 0
.IP "\fB\-\-eat\fR" 4
.IX Item "--eat"
.PD
Displays the contents of the Export Address Table section.
.IP "\fB\-I\fR" 4
.IX Item "-I"
.PD 0
.IP "\fB\-\-iat\fR" 4
.IX Item "--iat"
.PD
Displays the contents of the Import Address Table section.
.IP "\fB\-m\fR" 4
.IX Item "-m"
.PD 0
.IP "\fB\-\-runtime\fR" 4
.IX Item "--runtime"
.PD
Displays the contents of the runtine section.
.IP "\fB\-g\fR" 4
.IX Item "-g"
.PD 0
.IP "\fB\-\-debug\fR" 4
.IX Item "--debug"
.PD
Displays the contents of the debug information section.
.IP "\fB\-f\fR" 4
.IX Item "-f"
.PD 0
.IP "\fB\-\-config\fR" 4
.IX Item "--config"
.PD
Displays the contents of the configuration section.
.IP "\fB\-o\fR" 4
.IX Item "-o"
.PD 0
.IP "\fB\-\-resource\fR" 4
.IX Item "--resource"
.PD
Displays the contents of the resource information section.
.IP "\fB\-d\fR" 4
.IX Item "-d"
.PD 0
.IP "\fB\-\-detect\fR" 4
.IX Item "--detect"
.PD
Displays the contents of the signature analysis.
.IP "\fB\-x <number or name>\fR" 4
.IX Item "-x <number or name>"
.PD 0
.IP "\fB\-\-hex\-dump=<number or name>\fR" 4
.IX Item "--hex-dump=<number or name>"
.PD
Displays the contents of the indicated section as a hexadecimal bytes.
.IP "\fB\-c <number or name>\fR" 4
.IX Item "-c <number or name>"
.PD 0
.IP "\fB\-\-code\-dump=<number or name>\fR" 4
.IX Item "--code-dump=<number or name>"
.PD
Displays the contents of the indicated section as C Language Shell Code.
.IP "\fB\-s <number or name>\fR" 4
.IX Item "-s <number or name>"
.PD 0
.IP "\fB\-\-string\-dump=<number or name>\fR" 4
.IX Item "--string-dump=<number or name>"
.PD
Displays the contents of the indicated section as printable strings.
.IP "\fB\-u <number or name>\fR" 4
.IX Item "-u <number or name>"
.PD 0
.IP "\fB\-\-unicode\-dump=<number or name>\fR" 4
.IX Item "--unicode-dump=<number or name>"
.PD
Displays the contents of the indicated section as printable strings.
.IP "\fB\-r <number or name>\fR" 4
.IX Item "-r <number or name>"
.PD 0
.IP "\fB\-\-relocated\-dump=<number or name>\fR" 4
.IX Item "--relocated-dump=<number or name>"
.PD
Displays the contents of the indicated section as a hexadecimal bytes.
.IP "\fB\-i\fR" 4
.IX Item "-i"
.PD 0
.IP "\fB\-\-info\fR" 4
.IX Item "--info"
.PD
Equivalent to specifying: \fB\-\-hash\fR and \fB\-\-entropy\fR.
.IP "\fB\-X\fR" 4
.IX Item "-X"
.PD 0
.IP "\fB\-\-hash\fR" 4
.IX Item "--hash"
.PD
Calculates and displays the SHA-256 hash.
.IP "\fB\-e\fR" 4
.IX Item "-e"
.PD 0
.IP "\fB\-\-entropy\fR" 4
.IX Item "--entropy"
.PD
Displays the entropy of the object.
.RE
.IP "\fB\-C <number or name>\fR" 4
.IX Item "-C <number or name>"
.PD 0
.IP "\fB\-\-disassemble=<number or name>\fR" 4
.IX Item "--disassemble=<number or name>"
.PD
Display the assembler mnemonics for the machine instructions from the input file.
.IP "\fB\-M <options>\fR" 4
.IX Item "-M <options>"
.PD 0
.IP "\fB\-\-disassembler\-options=<options>\fR" 4
.IX Item "--disassembler-options=<options>"
.PD
Pass target specific information to the disassembler.
.RS 4
.ie n .IP """att""" 4
.IX Item "att"
.PD
Select the AT&T syntax mode.
.ie n .IP """intel""" 4
.IX Item "intel"
.PD
Select the Intel syntax mode.
.ie n .IP """arm32""" 4
.IX Item "arm32"
.PD 0
.ie n .IP """arm64""" 4
.IX Item "arm64"
.ie n .IP """i386""" 4
.IX Item "i386"
.ie n .IP """i8086""" 4
.IX Item "i8086"
.ie n .IP """x86-64""" 4
.IX Item "x86-64"
.ie n .IP """riscv32""" 4
.IX Item "riscv32"
.ie n .IP """riscv64""" 4
.IX Item "riscv64"
.ie n .IP """mips1""" 4
.IX Item "mips1"
.ie n .IP """mips2""" 4
.IX Item "mips2"
.ie n .IP """mips3""" 4
.IX Item "mips3"
.ie n .IP """mips4""" 4
.IX Item "mips4"
.ie n .IP """mips5""" 4
.IX Item "mips5"
.ie n .IP """mips32""" 4
.IX Item "mips32"
.ie n .IP """mips32r2""" 4
.IX Item "mips32r2"
.ie n .IP """mips32r6""" 4
.IX Item "mips32r6"
.ie n .IP """mips64""" 4
.IX Item "mips64"
.ie n .IP """mips64r2""" 4
.IX Item "mips64r2"
.ie n .IP """mips64r6""" 4
.IX Item "mips64r6"
.PD
Select disassembly for the given architecture.
.ie n .IP """amd64""" 4
.IX Item "amd64"
.PD 0
.ie n .IP """intel64""" 4
.IX Item "intel64"
.PD
Select between AMD64 ISA and Intel64 ISA.
.ie n .IP """addr16""" 4
.IX Item "addr16"
.PD 0
.ie n .IP """addr32""" 4
.IX Item "addr32"
.ie n .IP """addr64""" 4
.IX Item "addr64"
.ie n .IP """data16""" 4
.IX Item "data16"
.ie n .IP """data32""" 4
.IX Item "data32"
.PD
Specify the default address size and operand size.
.RE
.IP "\fB\-t <options>\fR" 4
.IX Item "-t <options>"
.PD 0
.IP "\fB\-\-script=<options>\fR" 4
.IX Item "--script=<options>"
.PD
Generate simple scripts to modify any section to reverse obfuscation.
.RS 4
.ie n .IP """add8""" 4
.IX Item "add8"
.PD 0
.ie n .IP """add16""" 4
.IX Item "add16"
.ie n .IP """add32""" 4
.IX Item "add32"
.PD
ADD (addition) the input with the given key.
.ie n .IP """not8""" 4
.IX Item "not8"
.PD 0
.ie n .IP """not16""" 4
.IX Item "not16"
.ie n .IP """not32""" 4
.IX Item "not32"
.PD
NOT (negation) the input.
.ie n .IP """rol8""" 4
.IX Item "rol8"
.PD 0
.ie n .IP """rol16""" 4
.IX Item "rol16"
.ie n .IP """rol32""" 4
.IX Item "rol32"
.PD
ROL (rotate-left) the input with the given key.
.ie n .IP """ror8""" 4
.IX Item "ror8"
.PD 0
.ie n .IP """ror16""" 4
.IX Item "ror16"
.ie n .IP """ror32""" 4
.IX Item "ror32"
.PD
ROR (rotate-right) the input with the given key.
.ie n .IP """rot5""" 4
.IX Item "rot5"
.PD 0
.ie n .IP """rot13""" 4
.IX Item "rot13"
.ie n .IP """rot18""" 4
.IX Item "rot18"
.ie n .IP """rot47""" 4
.IX Item "rot47"
.PD
Caesar substitution cipher which rotates alphabet characters by the specified amount.
.ie n .IP """shl8""" 4
.IX Item "shl8"
.PD 0
.ie n .IP """shl16""" 4
.IX Item "shl16"
.ie n .IP """shl32""" 4
.IX Item "shl32"
.PD
SHL (left-shift) the input with the given key.
.ie n .IP """shr8""" 4
.IX Item "shr8"
.PD 0
.ie n .IP """shr16""" 4
.IX Item "shr16"
.ie n .IP """shr32""" 4
.IX Item "shr32"
.PD
SHR (right-shift) the input with the given key.
.ie n .IP """sub8""" 4
.IX Item "sub8"
.PD 0
.ie n .IP """sub16""" 4
.IX Item "sub16"
.ie n .IP """sub32""" 4
.IX Item "sub32"
.PD
SUB (subtract) the input with the given key.
.ie n .IP """xor8""" 4
.IX Item "xor8"
.PD 0
.ie n .IP """xor16""" 4
.IX Item "xor16"
.ie n .IP """xor32""" 4
.IX Item "xor32"
.PD
XOR (eXclusive OR) the input with the given key.
.ie n .IP """dec""" 4
.IX Item "dec"
.PD 0
.ie n .IP """inc""" 4
.IX Item "inc"
.PD
Amount to increment or decrement the given key.

.SH SEE ALSO
detect-ng(1), convert-ng(1), coredump-ng(1), objcopy-ng(1), objdump-ng(1), objdwarf-ng(1), objhash-ng(1), readelf-ng(1), readpe-ng(1)

.SH COPYRIGHT
.IX Header "COPYRIGHT"
MIT License
.PP
Copyright (c) 2025 Kyle Wickens

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG
v1.2
-----------------------------------------------------------------------
[+] Added big-endian support.
[+] Added MIPS architecture support.
[+] Improved RISC-V architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################

v1.1
-----------------------------------------------------------------------
[+] Added ARM Thumb architecture support.
[+] Added RISC-V architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################

v1.0
-----------------------------------------------------------------------
[+] Added ARM architecture support.
[+] Improved x86-64 architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################



src/readpe-ng.c
#include "buffer.h"
#include "printf.h"
#include "readpe.h"
#include "options.h"
#include "scripts.h"
#include "objutils.h"

#include "static/readpe-ng.ci"

#define THIS_NAME "readpe-ng"

static int usage(poptions_t o, const char* name, const args_t args[], const int ecode) {
  int n = 0;
  n += usage_name(o, name, args, zDESCRIPTIONPE);
  n += usage_synopsis0(o, name, args);
  n += usage_synopsis2(o, name, zDISASSEMBLEARGS, zREADPEARGS2, zREADPEARGS3);
  n += usage_objname(o, name, "pefile");
  n += printf_eol();
  n += usage_description(o, name, args, zPEDESCRIPTION);
  n += usage_options0(o, name, args);
  n += usage_options2(o, name, zDISASSEMBLEARGS, zREADPEARGS2, zREADPEARGS3);
  n += usage_options2(o, name, zSCRIPTCOMMANDS, zREADPEARGS4, zREADPEARGS5);
  n += usage_seealso(o, name, args);
  n += usage_copyright(o, name, args);

  return ecode;
}

static int get_options_readpe(poptions_t o, int argc, char** argv, char* name) {
  if (0 == argc) {
    return usage(o, THIS_NAME, zREADPEARGS, ECODE_ARGUMENTS);
  }

  strname(o->prgname, name);
  printf_errname(o->prgname);

  for (int i = 0; i < argc; ++i) {
    if ('-' == argv[i][0] && '-' == argv[i][1]) {
      MALLOCA(char, arg0, 1024);
      MALLOCA(char, arg1, 1024);

      if (ECODE_ISOK(breakup_args(argv[i], arg0, NELEMENTS(arg0), arg1, NELEMENTS(arg1)))) {
        if (0 == xstrcmp(arg0, "--hex-dump")) {
          oinsertsname(o, ACT_HEXDUMP, arg1);
        } else if (0 == xstrcmp(arg0, "--string-dump")) {
          oinsertsname(o, ACT_STRDUMP8, arg1);
        } else if (0 == xstrcmp(arg0, "--unicode-dump")) {
          oinsertsname(o, ACT_STRDUMP16, arg1);
        } else if (0 == xstrcmp(arg0, "--relocated-dump")) {
          oinsertsname(o, ACT_RELDUMP, arg1);
        } else if (0 == xstrcmp(arg0, "--code-dump")) {
          oinsertsname(o, ACT_CODEDUMP, arg1);
        } else if (0 == xstrcmp(arg0, "--disassemble")) {
          oinsertsname(o, ACT_DISASSEMBLE, arg1);
        } else if (0 == xstrcmp(arg0, "--decompress")) {
          oinsertsname(o, ACT_ZLIB, arg1);
        } else if (0 == xstrcmp(arg0, "--script")) {
          sinsert(o, arg1, zSCRIPTCOMMANDS, NULL);
        }
      } else {
        imode_t action = get_options2(o, zREADPEARGS, argv[i]);
        if (0 == action) {
          return odeath(o, THIS_NAME, argv[i] + 2);
        }
        o->action |= action;
      }
    } else if ('-' == argv[i][0]) {
      if (0 == xstrcmp(argv[i], "-x")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_HEXDUMP, argv[++i]);
      } else if (0 == xstrcmp(argv[i], "-s")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_STRDUMP8, argv[++i]);
      } else if (0 == xstrcmp(argv[i], "-u")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_STRDUMP16, argv[++i]);
      } else if (0 == xstrcmp(argv[i], "-r")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_RELDUMP, argv[++i]);
      } else if (0 == xstrcmp(argv[i], "-c")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_CODEDUMP, argv[++i]);
      } else if (0 == xstrcmp(argv[i], "-C")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_DISASSEMBLE, argv[++i]);
      } else if (0 == xstrcmp(argv[i], "-z")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_ZLIB, argv[++i]);
      } else if (0 == xstrcmp(argv[i], "-t")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        sinsert(o, argv[++i], zSCRIPTCOMMANDS, NULL);
      } else {
        imode_t action = get_options1(o, zREADPEARGS, argv[i]);
        if (0 == action) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        o->action |= action;
      }
    } else if (0 == o->inpname[0]) {
      xstrncpy(o->inpname, argv[i], NELEMENTS(o->inpname));
    } else {
      return odeath(o, THIS_NAME, argv[i]);
    }
  }

  if (o->action & OPTPROGRAM_VERSION) {
    return version0(o, THIS_NAME, zREADPEARGS);
  }

  if (o->action & OPTPROGRAM_HELP) {
    return usage(o, THIS_NAME, zREADPEARGS, ECODE_OK);
  }

  return ECODE_OK;
}

int main(int argc, char* argv[]) {
  int x = ECODE_MALLOC;
  poptions_t o = omalloc();
  if (o) {
    x = get_options_readpe(o, argc - 1, argv + 1, argv[0]);
    if (ECODE_ISOK(x) && o->inpname[0]) {
      pbuffer_t p = bopen(o->inpname);
      if (p) {
        if (isPE(p)) {
          x = readpe(p, o);
        } else {
          printf_e("'%s': invalid file format.", o->inpname);
        }
      } else {
        printf_e("'%s': no such file.", o->inpname);
      }

      bfree(p);
    }

    ofree(o);
  }

  return x;
}

